### PR TITLE
Add 'continue' and 'tries' wget-like options to meterpreter 'download'

### PR DESCRIPTION
**Pull request by cayee**
When meterpreter download fails it is pretty inconvenient to download the whole file again. This pull request adds '-c' and ~~'-t'~~ '-l' options to meterpreter 'download' command (equivalent to '-c' and '--tries' options of wget) to continue the download from the place it got interrupted and to let it try repetitively when some network issues interrupt the download.
The options apply to file as well as to directory downloads including the recursive one
## Verification

Assuming a working meterpreter session:

I. '-c' option
-c
- [x] Start downloading a file
- [x] Interrupt it by pressing Ctrl+C
- [x] Check the size of the file downloaded

I.a restart with '-c' option
- [x] Restart downloading a file using -c option
- [x] Interrupt it by pressing Ctrl+C
- [x] Observe that the downloading part is bigger than at step 3.
- [x] Restart downloading a file using -c option
- [x] Observe that the size of the file being downloaded has not dropped
- [x] Once the download finishes, diff the file downloaded with the original one

I.b restart without '-c' - observe that the whole file is downloaded again
- [x] Restart downloading a file without additional options
- [x] Observe that the download starts from the beginning

II. ~~'-t'~~ '-l' option
-l
- [x] Start downloading a file with '-l 5'
- [x] Disconnect network from msf (i.e. disconnecting vm network adapter)
- [x] Observe msf trying to download the file again

II.a Network connection is restored
- [x] msf displays "resuming at " message
- [x] Observe size of file being downloaded - that the download does not start from the beginning
- [x] Once the download finishes, diff the file downloaded with the original one

II.b Network remains disconnected
- [x] After try number 5 the download finishes unsuccessfully

### CODE REVIEW

**Comment by ladyinthemiddle** [comment]
Awesome add-on!

File: lib/rex/post/meterpreter/extensions/stdapi/fs/dir.rb [hunk 1]
```diff
@@ -195,22 +195,51 @@ def Dir.unlink(path)
   # Downloads the contents of a remote directory a
   # local directory, optionally in a recursive fashion.
   #
-  def Dir.download(dst, src, recursive = false, force = true, glob = nil, &stat)
+  def Dir.download(dst, src, recursive = false, force = true, glob = nil, continue = false, tries = false, tries_no = 0, &stat)
+    tries_cnt = 0
+    begin
+      dir_files = self.entries(src)
```
Comment (OJ) [inline lib/rex/post/meterpreter/extensions/stdapi/fs/dir.rb]: Is the `glob` parameter omitted on purpose?

**Comment by OJ** [comment]
What is the intended and actual behaviour of the retry options when a directory is specified as the target of the download? If an interruption occurs during a single file, is the entire tree downloaded again? Is it smart enough to know that it doesn't need to redownload files that have succeeded during the first attempt?

**Comment by cayee** [comment]
@OJ I would rather say the fix is dumb enough to do what you ask :)
The new options are passed down to single file download that dir download uses, so when a particular file download is interrupted and resumed, dir download continues its work, so it does not redownload what it has already got.
Also an existing check skip downloading the file if the filename, size and mtime matches the one already downloaded

File: lib/rex/post/meterpreter/extensions/stdapi/fs/dir.rb (hunk 1 above)
Comment (cayee) [inline lib/rex/post/meterpreter/extensions/stdapi/fs/dir.rb]: Thanks for spotting. A fix coming up in a moment

**Comment by OJ** [comment]
Cool, cheers for the info :)

**Comment by OJ** [comment]
@cayee you get bonus points for signing your commits 👍

**Comment by wvu-r7** [comment]
Looks like this will need to be rebased because #7125 was landed. Sorry. :(

We should really use an options hash or similar (keyword arguments?) for the method parameters.

**Comment by cayee** [comment]
@wvu-r7 - added a hash for all download options. Changed '# tries' option to '-l'

**Comment by bcook-r7** [comment]
hmm:
```
meterpreter > download -c moodle-2.5.2.zip
[*] downloading: moodle-2.5.2.zip -> moodle-2.5.2.zip
[*] continuing from : 21037056 -> moodle-2.5.2.zip
^C[-] Error running command download: Interrupt
meterpreter > download -t 5 moodle-2.5.2.zip
[-] stdapi_fs_stat: Operation failed: The system cannot find the file specified.
```

**Comment by bcook-r7** [comment]
Hah, I forgot we had added '-t' for timestamp, and retries is '-l' now. Updated the description.

**Comment by bcook-r7** [comment]
Cool, looks good. Awesome work @cayee , thanks for the contribution.

**Comment by bcook-r7** [comment]
## Release Notes

The meterpreter 'download' command now has two new options. The first one, '-c' supports continuing a download for a file that was interrupted previously. The second one '-l <num>' specifies an automatic number of times meterpreter will attempt to download a file if it fails (due to a network interruption or other failure) before giving up. These options improve the resiliency of downloads and make getting large files more convenient.

File: lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/fs.rb [hunk 2]
```diff
@@ -397,7 +407,8 @@ def cmd_download(*args)
             src_path = file['path'] + client.fs.file.separator + file['name']
             dest_path = src_path.tr(src_separator, ::File::SEPARATOR)
 
-            client.fs.file.download(dest_path, src_path, ts) do |step, src, dst|
+            client.fs.file.download(dest_path, src_path, opts) do |step, src, dst|
+              puts step
```
Comment (wvu-r7) [inline lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/fs.rb]: Oops, looks like this was readded.

File: lib/rex/post/meterpreter/ui/console/command_dispatcher/extapi/clipboard.rb [hunk 3]
```diff
@@ -376,7 +376,8 @@ def download_file( dest_folder, source )
     dest = File.join( dest_folder, base )
 
     if stat.directory?
-      client.fs.dir.download( dest, source, true, true ) { |step, src, dst|
+      client.fs.dir.download( dest, source, {"recursive" => true}, true ) { |step, src, dst|
+#      client.fs.dir.download( dest, source, true, true ) { |step, src, dst|
```
Comment (wvu-r7) [inline lib/rex/post/meterpreter/ui/console/command_dispatcher/extapi/clipboard.rb]: This should be removed.

**Comment by bcook-r7** [comment]
See 022ab74f3077d6b1f87ea52073194f9ad8a4195c

**Pull request merged**